bindings/c/test/mako/ddsketch.hpp

```diff
@@ -0,0 +1,273 @@
+/*
+ * DDSketch.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2020 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef DDSKETCH_H
+#define DDSKETCH_H
+#include <limits>
+#include <type_traits>
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <vector>
+
+// A namespace for fast log() computation.
+namespace fastLogger {
+// Basically, the goal is to compute log(x)/log(r).
+// For double, it is represented as 2^e*(1+s) (0<=s<1), so our goal becomes
+// e*log(2)/log(r)*log(1+s), and we approximate log(1+s) with a cubic function.
+// See more details on Datadog's paper, or CubicallyInterpolatedMapping.java in
+// https://github.com/DataDog/sketches-java/
+inline const double correctingFactor = 1.00988652862227438516; // = 7 / (10 * log(2));
+constexpr inline const double A = 6.0 / 35.0, B = -3.0 / 5.0, C = 10.0 / 7.0;
+
+inline double fastlog(double value) {
+	int e;
+	double s = frexp(value, &e);
+	s = s * 2 - 1;
+	return ((A * s + B) * s + C) * s + e - 1;
+}
+
+inline double reverseLog(double index) {
+	long exponent = floor(index);
+	// Derived from Cardano's formula
+	double d0 = B * B - 3 * A * C;
+	double d1 = 2 * B * B * B - 9 * A * B * C - 27 * A * A * (index - exponent);
+	double p = cbrt((d1 - sqrt(d1 * d1 - 4 * d0 * d0 * d0)) / 2);
+	double significandPlusOne = -(B + p + d0 / p) / (3 * A) + 1;
+	return ldexp(significandPlusOne / 2, exponent + 1);
+}
+} // namespace fastLogger
+
+// DDSketch for non-negative numbers (those < EPS = 10^-18 are
+// treated as 0, and huge numbers (>1/EPS) fail ASSERT). This is the base
+// class without a concrete log() implementation.
+template <class Impl, class T>
```

Comment:
Not sure if CRTP is warranted here.
Looks like `Impl`s are mostly stateless and its relationship to `DDSketchBase` is clearly not a `is-a` relationship.
`Impl`s are more of a set of plug-in constant and functions to `DDSketchBase`.
Could we remove the inheritance, make it e.g. `DDSketch<Plugin>` and use `has-a` relationship instead? Looks like `T`, `EPS`, `defaultMin()`, `defaultMax()`, etc. could all be `Plugin`'s nested traits. Instead of having derived class invoke `setBucketSize()`, `DDSketch` could be calling `buckets.resize(plugin.getBucketSize())`. `T` could be `Plugin::SampleType`, and so on.
That would leave `DDSketch` to deal purely with the abstract sample accounting and defer everything value-specific to the `Plugin`.

Reply:
So the DDSketch code comes primarily from this PR:
https://github.com/apple/foundationdb/pull/4088
The idea is that this will eventually be merged into fdb, and mako can use that implementation instead of the one in this change. However, instead of waiting for the PR to be merged in we decided to copy the implementation into mako for now. I think making design changes to DDSketch is beyond the scope of this PR. I followed your recommendations on other changes, such as using `static` for stateless methods and using iterators for the backwards loop.